Tests/Integration/DeduplicationStrategyIntegrationTests.cs

```diff
@@ -26,111 +32,143 @@ public DeduplicationStrategyIntegrationTests(ITestOutputHelper testOutputHelper)
         public async Task Validate_Dup_Message_Processing(DeduplicationMode mode)
         {
             var simulatedDevice = new SimulatedDevice(TestDeviceInfo.CreateABPDevice(1));
-            var messageProcessed = mode == DeduplicationMode.Drop;
-            messageProcessed = false;
-
-            LoRaDeviceApi.Setup(x => x.ExecuteFunctionBundlerAsync(simulatedDevice.DevEUI, It.IsNotNull<FunctionBundlerRequest>()))
-                .Returns((DevEui _, FunctionBundlerRequest _) =>
-                {
-                    var isDup = messageProcessed;
-                    messageProcessed = true;
-                    return Task.FromResult(new FunctionBundlerResult()
-                    {
-                        DeduplicationResult = new DeduplicationResult
-                        {
-                            IsDuplicate = isDup
-                        }
-                    });
-                });
-
-            LoRaDeviceApi
-                .Setup(x => x.NextFCntDownAsync(It.IsAny<DevEui>(), It.IsAny<uint>(), It.IsAny<uint>(), It.IsAny<string>()))
-                .ReturnsAsync(simulatedDevice.FrmCntDown + 1)
-                .Callback(() =>
-                {
-                    // this call should only be made, if we do not have a deduplication strategy
-                    // since otherwise we expect the fcntDown to be calculated in the same API call as deduplication
-                    Assert.True(mode == DeduplicationMode.None);
-                });
-
-            LoRaDeviceApi.Setup(x => x.ABPFcntCacheResetAsync(It.IsNotNull<DevEui>(), It.IsAny<uint>(), It.IsNotNull<string>()))
-                .ReturnsAsync(true);
+            var messageProcessed = false;
 
-            var shouldBeMarked = false;
+            foreach (var api in new[] { LoRaDeviceApi, SecondLoRaDeviceApi })
+            {
+                api.Setup(x => x.ExecuteFunctionBundlerAsync(simulatedDevice.DevEUI, It.IsNotNull<FunctionBundlerRequest>()))
+                   .ReturnsAsync((DevEui _, FunctionBundlerRequest _) =>
+                   {
+                       lock (this.functionLock)
+                       {
+                           var isDup = messageProcessed;
+                           messageProcessed = true;
+                           return new FunctionBundlerResult()
+                           {
+                               DeduplicationResult = new DeduplicationResult { IsDuplicate = isDup }
+                           };
+                       }
+                   });
+
+                api.Setup(x => x.NextFCntDownAsync(It.IsAny<DevEui>(), It.IsAny<uint>(), It.IsAny<uint>(), It.IsAny<string>()))
```

Comment:
I don't think we need this setup in any of the modes since we are sending unconfirmed messages.

Reply:
Good catch, refactored the test further in 734c7ae8f3edc7a3ee3effa347fb797ac1a00bb7